doc/source/devel/array.rst

```diff
@@ -0,0 +1,612 @@
+``[[-100,2],[-6,-15]]`` of the full array,
+``sub_of_sub`` represents ``[[-100],[-6]]``, and ``sub2`` represents
+``[[3,-6,-15]]`` (at this point the ``data`` pointer holds
+``[0, -100, 2, 3, -6, -15]``).
+
+The fact that ``arr`` originally wrapped ``data`` has no bearing on
+the outcomes described above for each instance of ``CelloArray``.
+We illustrate this below with an analogous abreviated example, where
+the analog to ``arr``, called ``array``, originally owns its data.
+
+.. code-block:: c++
+
+   CelloArray<int,2> array(2,3);
+   array(0,0) = 0;    array(0,1) = 1;    array(0,2) = 2;
+   array(1,0) = 3;    array(1,1) = 4;    array(1,2) = 5;
+   CelloArray<int,2> subarray = array.subarray(CSlice(0,2), CSlice(1,3));
+   array(1,3) *= -3;
+   subarray(0,0) = -100;
+   CelloArray<int,2> subarray_of_subarray = subarray.subarray(CSlice(0,2),
+                                                              CSlice(0,1));
+   subarray_of_subarray(1,0) += 8;
+
+After executing the preceeding block of code, ``array`` reflects
+``[[0,-100,2],[3,12,-15]]``, while ``subarray`` and
+``subarray_of_subarray`` represent the subarrays
+``[[-100,2],[12,-15]]`` and ``[[-100],[12]]``. If this was all the
+code we executed, the memory of ``array`` would be freed after its
+destructor and the destructors of all of subarrays or shallowcopies
+are called.
+
+If we reassign ``array`` to a different array, just like before, the values
+of its subarrays and shallow copies will be unaffected.
+
+.. code-block:: c++
+
+   CelloArray<int,2> subarray2 = array.subarray(CSlice(1,2),CSlice(0,3));
+   array = CelloArray<int, 2>(3,3);
+   subarray(1,0) /= -2;
+
+Now, ``subarray`` represents ``[[-100,2],[-6,-15]]`` from the full
+array, ``subarray_of_subarray`` represents ``[[-100],[-6]]``, and
+``subarray2`` represents ``[[3,-6,-15]]``. We note that no memory
+has been deallocated. The memory will only be deallocated after
+``subarray``, ``subarray_of_subarray``, and ``subarray2`` have
+all had their deconstructor called and/or been assigned unrelated
+arrays, assuming no additional subarrays or shallowcopies of any of
+the 3 variables are made in the meantime (in that case the memory
+would still not be deallocated until any additional
+subarrays/shallowcopies that view the original data are destroyed).
+
+Additional CSlice features
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+``CSlice`` provides two additional features to simplify code when
+the generating subarrays of a ``CelloArray`` instance. These are
+
+  1. The constructor supports negative indexing. For example
+     ``CSlice(1,-1)`` represents a slice starting at the second
+     element and stopping at (does not include) the last element
+     along a dimension. Additionally, ``CSlice(-3,-1)`` represents
+     starting from the third-to-last and stopping at the last
+     element along a given dimension.
+  2. The constructor accepts the ``NULL`` and ``nullptr`` as the
```

Comment:
I could imaging that allowing `NULL` may interfere with using `0` as index here.
The functionality is great, but I think it could be more specific, e.g., introducing an `sl::ALL`, `sl::BEG`, and `sl::END` (or something along those line).

Reply:
Forrest and I were having a conversation about this above. Basically, it just means that `CSlice(3,NULL)` and `CSlice(3,0)` are equivalent.

Since both of you suggested building this into the time system, I'll adapt to your idea